File: app/repositories/tax_rate_repository.py
import uuid
from decimal import Decimal

from sqlalchemy import select, update
from sqlalchemy.orm import Session

from app.models.tax_rate import TaxRate


def get(db: Session, tenant_id: uuid.UUID, tax_rate_id: uuid.UUID) -> TaxRate | None:
    stmt = select(TaxRate).where(TaxRate.tenant_id == tenant_id, TaxRate.id == tax_rate_id)
    return db.scalars(stmt).first()


def get_default(db: Session, tenant_id: uuid.UUID) -> TaxRate | None:
    stmt = select(TaxRate).where(TaxRate.tenant_id == tenant_id, TaxRate.is_default.is_(True))
    return db.scalars(stmt).first()


def list_for_tenant(db: Session, tenant_id: uuid.UUID) -> list[TaxRate]:
    stmt = select(TaxRate).where(TaxRate.tenant_id == tenant_id).order_by(TaxRate.name)
    return list(db.scalars(stmt))


def clear_default(db: Session, tenant_id: uuid.UUID) -> None:
    """Hay un indice unico parcial de un solo default por tenant: sin limpiar
    el anterior, marcar uno nuevo revienta contra la base."""
    db.execute(
        update(TaxRate)
        .where(TaxRate.tenant_id == tenant_id, TaxRate.is_default.is_(True))
        .values(is_default=False)
    )
    db.flush()


def create(
    db: Session,
    *,
    tenant_id: uuid.UUID,
    name: str,
    rate: Decimal,
    included_in_price: bool,
    is_default: bool,
) -> TaxRate:
    tax_rate = TaxRate(
        tenant_id=tenant_id,
        name=name,
        rate=rate,
        included_in_price=included_in_price,
        is_default=is_default,
    )
    db.add(tax_rate)
    db.flush()
    return tax_rate

File: php/src/Repositories/TaxRateRepository.php
<?php

declare(strict_types=1);

namespace App\Repositories;

use App\Core\Row;
use App\Models\TaxRate;
use PDO;

final class TaxRateRepository
{
    public function __construct(private readonly PDO $pdo)
    {
    }

    public function get(string $tenantId, string $taxRateId): ?TaxRate
    {
        $stmt = $this->pdo->prepare('SELECT * FROM tax_rates WHERE tenant_id = :tenant_id AND id = :id');
        $stmt->execute(['tenant_id' => $tenantId, 'id' => $taxRateId]);
        $row = $stmt->fetch();
        return $row === false ? null : TaxRate::fromRow($row);
    }

    public function getDefault(string $tenantId): ?TaxRate
    {
        $stmt = $this->pdo->prepare('SELECT * FROM tax_rates WHERE tenant_id = :tenant_id AND is_default = true');
        $stmt->execute(['tenant_id' => $tenantId]);
        $row = $stmt->fetch();
        return $row === false ? null : TaxRate::fromRow($row);
    }

    /** @return TaxRate[] */
    public function listForTenant(string $tenantId): array
    {
        $stmt = $this->pdo->prepare('SELECT * FROM tax_rates WHERE tenant_id = :tenant_id ORDER BY name');
        $stmt->execute(['tenant_id' => $tenantId]);
        return array_map(TaxRate::fromRow(...), $stmt->fetchAll());
    }

    /**
     * Hay un indice unico parcial de un solo default por tenant: sin limpiar
     * el anterior, marcar uno nuevo revienta contra la base.
     */
    public function clearDefault(string $tenantId): void
    {
        $stmt = $this->pdo->prepare(
            'UPDATE tax_rates SET is_default = false WHERE tenant_id = :tenant_id AND is_default = true'
        );
        $stmt->execute(['tenant_id' => $tenantId]);
    }

    public function create(
        string $tenantId,
        string $name,
        string $rate,
        bool $includedInPrice,
        bool $isDefault,
    ): TaxRate {
        $stmt = $this->pdo->prepare(
            'INSERT INTO tax_rates (tenant_id, name, rate, included_in_price, is_default)
             VALUES (:tenant_id, :name, :rate, :included_in_price, :is_default)
             RETURNING *'
        );
        $stmt->execute([
            'tenant_id' => $tenantId,
            'name' => $name,
            'rate' => $rate,
            'included_in_price' => Row::pgBool($includedInPrice),
            'is_default' => Row::pgBool($isDefault),
        ]);
        return TaxRate::fromRow($stmt->fetch());
    }

    public function setDefault(string $tenantId, string $taxRateId): TaxRate
    {
        $stmt = $this->pdo->prepare(
            'UPDATE tax_rates SET is_default = true WHERE tenant_id = :tenant_id AND id = :id RETURNING *'
        );
        $stmt->execute(['tenant_id' => $tenantId, 'id' => $taxRateId]);
        return TaxRate::fromRow($stmt->fetch());
    }
}
